streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -628,11 +635,40 @@ public StreamThread(final Time time,
             new InternalConsumerConfig(config.getMainConsumerConfigs("dummyGroupId", "dummyClientId", dummyThreadIdx))
                 .getInt(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG);
         this.commitTimeMs = config.getLong(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG);
-        this.eosBetaEnabled = "exactly_once_beta".equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG));
+        this.eosBetaEnabled = eosBetaEnabled(config);
+        this.eosUpgradeModeEnabled = eosUpgradeModeEnabled(config);
 
         this.numIterations = 1;
     }
 
+    static boolean eosAlphaEnabled(final StreamsConfig config) {
+        return StreamsConfig.EXACTLY_ONCE.equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG));
+    }
+
+    public static boolean eosBetaEnabled(final StreamsConfig config) {
+        return "exactly_once_beta".equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG));
+    }
+
+    public static boolean eosEnabled(final StreamsConfig config) {
+        return eosAlphaEnabled(config) || eosBetaEnabled(config);
+    }
+
+    public static boolean eosUpgradeModeEnabled(final StreamsConfig config) {
```

Comment:
q: I thought we don't need to use upgrade from mode for 2.5 anymore? Just the `beta` should be sufficient?

Reply:
If you want to enable eos-beta, you need to do two rolling bounces: in the first you set eos-beta plus upgrade.from: for this case, we stay at producer-per-task model, but we already pass consumer-group-metadata into the producer on tx-commit (ie, we enable consumer-side fencing). Afterwards, you do a second rolling bounce removing upgrade.from config and we will switch to eos-beta with producer-per-thread.

If we don't do two rolling bounces, and don't have proper fencing during the upgrade phase when some instances use eos-alpha and some use eos-beta.